### PR TITLE
Define route, endpoint and service for kubernetes

### LINKED ISSUE #283: [Kb Access 3] Define route, endpoint, and service
# Define route, api endpoint, and service

In `config.ts`:
- [x] Define a kubernetes `route` and `api endpoint` and its getters. Since kubernetes viewing is cluster specific, remember to pass in the `clusterId` to this route and endpoint.

`services/kubernetes/` is a directory that contains all logic for retrieving/processing/typing response values from api:
- [x] Create a kubernetes class that will define a fetch method to get a list of all kube servers for the specified clusterId.
- [x] The response value from the fetch method (or any api responses really) will need to be processed. The convention is to create a separate file named `make.ts`, and it will contain methods named `makeXXX` that will be responsible for processing the responses we get from api. There are variety of reasons why we want to process the response values:
   1. Arrays may come as `undefined` and needs to be replaced with an empty array to avoid accessing undefined props errors. 
   1. Setting default values to objects if undefined
   1. Sometimes json response contains field names containing underscore ie. `expires_in`. JS conventions dictates we use camelCase, so we can rename this field to `expiresIn: json.expires_in`
   1. Sometimes with the response given, we want to add an extra field. For example, if the response value contained a field `date: Date`, and we wanted a formatted string version of this date, we can add a field `dateText: stringValue` (while keeping the original data). These text values often helps with table searching.
   1. All of the above helps create consistent responses sent to our components
- [x] Instantiate this class in our `teleportContext.tsx` file, so that children components can have access to it.
- [x] Test the class you made, to ensure the response received from api, is processed and formatted as you expected.

### Mocked response
This is what response data would look like from the api
```tsx
[
  {
    name: 'tele.logicoma.dev-prod',
    labels: [
      { name: 'kernal', value: '4.15.0-51-generic' },
      { name: 'env', value: 'prod' },
    ],
  },
]
````

This is what data should look like after being processed:
```tsx
[
  {
    name: 'tele.logicoma.dev-prod',
    labels: [ 'kernal: 4.15.0-51-generic', 'env: prod' ],
  },
]
````

### Note
We won't be using this service until the next issue.

### PR DESCRIPTION
**Pull request by amha19**
**Description**

- Defines a kubernetes route and api endpoint that takes in a cluster id.
- It defines getter method for the route and the api.
- Creates a fetch method and fetch response processing method.

Given this:
```
[
  {
    name: 'tele.logicoma.dev-prod',
    tags: [
      { name: 'kernal', value: '4.15.0-51-generic' },
      { name: 'env', value: 'prod' },
    ],
  },
]
```

Will output this:

```
[
  {
    name: 'tele.logicoma.dev-prod',
    tags: [ 'kernal: 4.15.0-51-generic', 'env: prod' ],
  },
]
```

And it adds this service to the teleport context.

This PR closes #283

### CODE REVIEW

File: packages/teleport/src/services/kube/kubernetes.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { map } from 'lodash';
+import api from 'teleport/services/api';
+import cfg from 'teleport/config';
+import makeKubernetes from './makeKubernetes';
+
+const service = {
```
Comment (kimlisa) [inline packages/teleport/src/services/kube/kubernetes.ts:22]: hmm... is this a class?

Comment (amha19) [reply packages/teleport/src/services/kube/kubernetes.ts]: No, it is not :disappointed:. I was following Nodes structure, I forgot it was supposed to be a class.

File: packages/teleport/src/services/kube/index.ts [hunk 2]
```diff
@@ -14,4 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+import KubernetesService from './kube';
```
Comment (kimlisa) [inline packages/teleport/src/services/kube/index.ts:17]: Lets rename to `KubeService`

File: packages/teleport/src/config.ts [hunk 3]
```diff
@@ -282,6 +288,10 @@ const cfg = {
     return generatePath(cfg.api.rolesPath, { name });
   },
 
+  getClusterKubernetesUrl(clusterId: string) {
```
Comment (kimlisa) [inline packages/teleport/src/config.ts:291]:
```suggestion
  getKubernetesUrl(clusterId: string) {
```

File: packages/teleport/src/services/kube/makeKube.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Kube } from './types';
+
+export default function makeKube(json): Kube {
+  const { name, tags = [] } = json;
```
[the remark this answers is not in the document]
Comment (amha19) [reply packages/teleport/src/services/kube/makeKube.ts]: I've now changed the coming json field from `tags` to `labels`.

File: packages/teleport/src/services/kube/kube.test.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import KubeService from './kube';
+import api from 'teleport/services/api';
+
```
Comment (kimlisa) [inline packages/teleport/src/services/kube/kube.test.ts:19]: Also add a test, if response comes back undefined. Arrays in particular may come as undefined.

Comment (kimlisa) [reply packages/teleport/src/services/kube/kube.test.ts:19]: if you want to know why: https://github.com/golang/go/issues/37711

File: packages/teleport/src/services/kube/makeKube.ts [hunk 6]
```diff
@@ -17,7 +17,8 @@ limitations under the License.
 import { Kube } from './types';
 
 export default function makeKube(json): Kube {
-  const { name, labels = [] } = json;
+  const { name } = json;
+  const labels = json.labels ? json.labels : [];
```
Comment (kimlisa) [inline packages/teleport/src/services/kube/makeKube.ts:21]:
```suggestion
  const labels = json.labels || [];
```

simplified. as stated earlier, `undefined` and `null` are falsey values, so if false, empty array will be assigned

File: packages/teleport/src/services/kube/kube.test.ts [hunk 7]
```diff
@@ -17,18 +17,43 @@ limitations under the License.
...
-  const kubeService = new KubeService();
-  const response = await kubeService.fetchKubernetes('clusterId');
-
-  expect(response).toEqual([
-    {
-      name: 'tele.logicoma.dev-prod',
-      tags: ['kernal: 4.15.0-51-generic', 'env: prod'],
-    },
-  ]);
+describe('correct processed fetch response', () => {
```
[the remark this answers is not in the document]
Comment (amha19) [reply packages/teleport/src/services/kube/kube.test.ts]: Good to know. Thanks!

**Review by kimlisa — APPROVED** [review]
LGTM

**Pull request merged**